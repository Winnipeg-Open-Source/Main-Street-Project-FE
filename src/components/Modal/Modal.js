import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { Flex, mediaQueries } from 'pcln-design-system';
import { Modal as PCLNModal } from 'pcln-modal';
import ModalHeader from 'components/ModalHeader';
import ModalFooter from 'components/ModalFooter';

const StyledModal = styled(PCLNModal)`
    max-width: 100%;
    height: 100%;

    & > div {
        & > :last-child {
            height: 100%;
            padding: 0;
        }
    }

    ${mediaQueries.sm} {
        max-width: calc(100% - 32px);
        height: auto;
    }
`;

const ContentWrapper = styled(Flex)`
    flex-direction: column;
    height: 100%;

    & > :first-child {
        overflow-y: scroll;
        flex: 1;
    }

    ${mediaQueries.sm} {
        min-height: 150px;
        max-height: 400px;
        height: auto;
    }
`;

/* istanbul ignore next */
const OVERLAY_ANIMATIONS = (state) => `
    opacity: 0;
    transition: opacity .25s ease-out;
    ${state === 'entering' ? `opacity: 0;` : ''}
    ${state === 'entered' ? `opacity: 1;` : ''}
    ${state === 'exiting' ? `opacity: 1;` : ''}
    ${state === 'exited' ? `opacity: 0;` : ''}
`;

/* istanbul ignore next */
const DIALOG_ANIMATIONS = (state) => `
    transform: translate(-100%);
    transition: transform .25s ease-out;
    ${state === 'entering' ? `transform: translate(-100%);` : ''}
    ${state === 'entered' ? `transform: translate(0);` : ''}
    ${state === 'exiting' ? `transform: translate(-100%);` : ''}
    ${state === 'exited' ? `transform: translate(-100%);` : ''}
`;

function Modal ({
    className,
    isMobile,
    isOpen,
    hasFooter,
    title,
    subtitle,
    children,
    onClose,
}) {
    return (
        <StyledModal
            className={className}
            disableCloseButton
            isOpen={isOpen}
            timeout={250}
            color='background.lightest'
            width={['100%', null, '640px']}
            header={
                <ModalHeader
                    title={title}
                    subtitle={subtitle}
                    onClose={onClose}
                />
            }
            overlayAnimation={isMobile ? OVERLAY_ANIMATIONS : null}
            dialogAnimation={isMobile ? DIALOG_ANIMATIONS : null}
            onClose={onClose}
        >
            <ContentWrapper p={3}>
                {children}
                {hasFooter && <ModalFooter onClose={onClose} />}
            </ContentWrapper>
        </StyledModal>
    );
}

Modal.displayName = 'Modal';

Modal.propTypes = {
    className: PropTypes.string,
    isMobile: PropTypes.bool,
    isOpen: PropTypes.bool,
    hasFooter: PropTypes.bool,
    title: PropTypes.string,
    subtitle: PropTypes.string,
    children: PropTypes.node,
    onClose: PropTypes.func,
};

Modal.defaultProps = {
    className: '',
};

export default Modal;
