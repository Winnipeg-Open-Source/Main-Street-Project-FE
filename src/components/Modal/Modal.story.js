import React from 'react';
import { action } from '@storybook/addon-actions';
import Modal from './Modal';

const Content = () => <div>Content</div>;

const LargeContent = () => (
    <div>
        <div>Paragraphs</div>
        <p>Paragraph 1</p>
        <p>Paragraph 2</p>
        <p>Paragraph 3</p>
        <p>Paragraph 4</p>
        <p>Paragraph 5</p>
        <p>Paragraph 6</p>
        <p>Paragraph 7</p>
        <p>Paragraph 8</p>
        <p>Paragraph 9</p>
        <p>Paragraph 10</p>
        <p>Paragraph 11</p>
        <p>Paragraph 12</p>
    </div>
);

export const TitleOnly = () => (
    <Modal isOpen title='Title'>
        <Content />
    </Modal>
);

export const WithSubtitle = () => (
    <Modal isOpen title='Title' subtitle='Subtitle Message'>
        <Content />
    </Modal>
);

export const WithCloseButton = () => (
    <Modal isOpen title='Title' subtitle='Subtitle Message' onClose={action('Close')}>
        <Content />
    </Modal>
);

export const WithFooter = () => (
    <Modal isOpen hasFooter title='Title' subtitle='Subtitle Message' onClose={action('Close')}>
        <Content />
    </Modal>
);

export const Scrollable = () => (
    <Modal isOpen title='Title' subtitle='Subtitle Message' onClose={action('Close')}>
        <LargeContent />
    </Modal>
);

export default {
    title: 'Modal',
    component: Modal,
};
