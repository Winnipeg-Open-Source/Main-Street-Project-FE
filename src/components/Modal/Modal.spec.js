import React, { useState } from 'react';
import { fireEvent, wait } from '@testing-library/react';
import renderWithTheme from 'tests/helpers/renderWithTheme';
import Modal from './Modal';

function ModalContainer (props) {
    const [isOpen, setOpen] = useState(false);
    const onOpenClick = () => setOpen(true);
    const onCloseClick = () => setOpen(false);

    return (
        <>
            <button onClick={onOpenClick}>Button</button>
            <Modal isOpen={isOpen} onClose={onCloseClick} {...props}>
                Content
            </Modal>
        </>
    );
}

describe('Modal', () => {
    it('renders without crashing', () => {
        const { asFragment } = renderWithTheme(<Modal isOpen />);

        expect(asFragment()).toMatchSnapshot();
    });

    it('should open when button clicked', () => {
        const { queryByText } = renderWithTheme(<ModalContainer />);

        expect(queryByText(/Content/)).not.toBeInTheDocument();
        fireEvent.click(queryByText(/Button/));
        expect(queryByText(/Content/)).toBeInTheDocument();
    });

    it('should close when close button clicked', async () => {
        const { queryByText, queryByTestId } = renderWithTheme(
            <ModalContainer />
        );

        expect(queryByText(/Content/)).not.toBeInTheDocument();
        fireEvent.click(queryByText(/Button/));
        expect(queryByText(/Content/)).toBeInTheDocument();
        fireEvent.click(queryByTestId('close-icon'));
        await wait(() =>
            expect(queryByText(/Content/)).not.toBeInTheDocument()
        );
    });

    it('has footer with close button', async () => {
        const { queryByText } = renderWithTheme(<ModalContainer hasFooter />);

        fireEvent.click(queryByText(/Button/));
        expect(queryByText(/Content/)).toBeInTheDocument();
        fireEvent.click(queryByText(/Close/));
        await wait(() =>
            expect(queryByText(/Content/)).not.toBeInTheDocument()
        );
    });
});
