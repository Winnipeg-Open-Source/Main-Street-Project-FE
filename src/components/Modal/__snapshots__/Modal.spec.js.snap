// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal renders without crashing 1`] = `
<DocumentFragment>
  <div
    class="sc-qOvoE hIfWnJ sc-qWSGh dpgLb"
  />
</DocumentFragment>
`;
